core/src/main/scala/kafka/server/ControllerApis.scala

```diff
@@ -538,4 +542,57 @@ class ControllerApis(val requestChannel: RequestChannel,
         }
       })
   }
+
+  def handleCreatePartitions(request: RequestChannel.Request): Unit = {
+    val responses = createPartitions(request.body[CreatePartitionsRequest].data,
+      authHelper.authorize(request.context, CREATE, CLUSTER, CLUSTER_NAME),
+      names => authHelper.filterByAuthorized(request.context, CREATE, TOPIC, names)(n => n))
+    requestHelper.sendResponseMaybeThrottle(request, throttleTimeMs => {
+      val responseData = new CreatePartitionsResponseData().
+        setResults(responses).
+        setThrottleTimeMs(throttleTimeMs)
+      new CreatePartitionsResponse(responseData)
+    })
+  }
+
+  def createPartitions(request: CreatePartitionsRequestData,
+                       hasClusterAuth: Boolean,
+                       getCreatableTopics: Iterable[String] => Set[String]): util.List[CreatePartitionsTopicResult] = {
+    val responses = new util.ArrayList[CreatePartitionsTopicResult]()
+    val duplicateTopicNames = new util.HashSet[String]()
+    val topicNames = new util.HashSet[String]()
+    request.topics().forEach {
+      topic =>
+        if (!topicNames.add(topic.name())) {
+          duplicateTopicNames.add(topic.name())
+        }
+    }
+    duplicateTopicNames.forEach {
+      case topicName => responses.add(new CreatePartitionsTopicResult().
```

Comment:
nit: the `case` is not needed. usually we put the argument on the previous line.
```scala
    duplicateTopicNames.forEach { topicName =>
```
Same in the `topicNames` loop below.